torchaudio/datasets/datasets.py

```diff
@@ -0,0 +1,419 @@
+import os
+import pickle
+import random
+from functools import partial, reduce
+from warnings import warn
+
+import torch
+
+import torchaudio
+
+
+class Cache:
+    """
+    Wrap a generator so that, whenever a new item is returned, it is saved to disk in a pickle.
+    """
+
+    def __init__(self, generator, location):
+        self.generator = generator
+        self.location = location
+
+        self._id = id(self)
+        self._cache = []
+        self._internal_index = 0
+
+    def __iter__(self):
+        self._internal_index = 0
+        return self
+
+    def __next__(self):
+        if self._internal_index < len(self):
+            item = self[self._internal_index]
+        else:
+            item = next(self.generator)
+
+            file = str(self._id) + "-" + str(len(self))
+            file = os.path.join(self.location, file)
+            self._cache.append(file)
+
+            os.makedirs(self.location, exist_ok=True)
+            with open(file, "wb") as file:
+                pickle.dump(item, file)
+
+        self._internal_index += 1
+        return item
+
+    def __getitem__(self, index):
+        file = self._cache[index]
+        with open(file, "rb") as file:
+            item = pickle.load(file)
+        return item
+
+    def __len__(self):
+        # Return length of cache
+        return len(self._cache)
+
+
+class Buffer:
+    """
+    Wrap a generator so as to keep the last few in memory.
+    """
+
+    def __init__(self, generator, capacity=10):
+        self.generator = generator
+        self.capacity = capacity
+        self._cache = []
+        self._fill()
+
+    def _fill(self):
+        while len(self._cache) <= self.capacity:
+            self._cache.append(next(self.generator))
+
+    def __getitem__(self, n):
+        self._fill()
+        return self._cache[n]
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        item = self._cache.pop(0)
+        self._fill()
+        return item
+
+
+def download(urls, root_path):
+    """
+    Download each url to root_path.
+
+    Input: url generator, folder inside archive
+    Output: downloaded archive, folder inside archive
+    """
+    for url, folder in urls:
+        torchaudio.datasets.utils.download_url(url, root_path)
+        file = os.path.join(root_path, os.path.basename(url))
+        yield file, folder
+
+
+def extract(files):
```

Comment:
You should add a flag to have the user choose a format and also I don't think you need to make this specific to files. Have a look at https://github.com/pytorch/text/blob/fd31bf3722e17dfd5998b8c4f32c0431f3016d59/torchtext/utils.py#L152 , which however, regrettably, still requires files.

For an example of how to extract "on-the-fly" from an arbitrary file / buffer check out this snippet

```
def get_data(URL):
    r = requests.get(URL)
    file_like_object = io.BytesIO(r.content)
    tar = tarfile.open(fileobj=file_like_object)
    d = {}
    for member in tar.getmembers():
        if member.isfile() and member.name.endswith('csv'):
            k = 'train' if 'train' in member.name else 'test'
            d[k] = tar.extractfile(member)
    return d```

With something generic like this users can do

```
for tar_buffer in extract(open("/path/to/file.tar")
    tar_buffer.write_to_disk("/path...")
```

In general https://docs.python.org/3/library/io.html is a good library to look at for this.

If you want to make it explicit that this extracts form and to a file, I'd change the name to extract_file_to_disk or such.